### PR TITLE
Poem 039 implementation

### LINKED ISSUE #1824: Implementation of POEM 039
### Summary of Issue

User function registration for ExecComps.

### Issue Type

- [ ] Bug
- [ x] Enhancement  
      POEM #039

### PR DESCRIPTION
**Pull request by naylor-b**
### Summary

Adds function registration and the shape_by_conn option to ExecComp.  Also, ExecComp can now use it's built-in 'cs' implementation for partials or it can use openmdao's 'fd' or 'cs' implementations.

- also added try-finally blocks to a bunch of our context managers

### Related Issues

- Resolves #1824

### Backwards incompatibilities

None

### New Dependencies

None

### CODE REVIEW

File: openmdao/components/tests/test_exec_comp.py [hunk 1]
```diff
@@ -1196,6 +1218,320 @@ def test_list_outputs_resids_tol(self):
...
+            om.ExecComp.register('area', lambda x: x**2, complex_safe=True)
+            p = om.Problem()
+            comp = p.model.add_subsystem('comp', om.ExecComp('area_square = area(x)', shape=size))
+            comp.declare_partials('area_square', 'x', method='cs')
+            p.setup()
+            p['comp.x'] = 3.
+            p.run_model()
+            assert_near_equal(p['comp.area_square'], np.ones(size) * 9., 1e-6)
+            J = p.compute_totals(of=['comp.area_square'], wrt=['comp.x'])
+            assert_near_equal(J['comp.area_square', 'comp.x'], np.eye(size) * 6., 1e-6)
```
Comment (Kenneth-T-Moore) [inline openmdao/components/tests/test_exec_comp.py:1383]: Might make sense to use a tighter tolerance on the 'cs' test, so you know it's not doing fd.

File: openmdao/components/exec_comp.py [hunk 2]
```diff
@@ -74,28 +75,10 @@ class ExecComp(ExplicitComponent):
         Default is None, which means units are provided for variables individually.
     complex_stepsize : double
         Step size used for complex step which is used for derivatives.
+    _manual_decl_partials : bool
+        If True, at least one partial has been declared by the user.
```
Comment (Kenneth-T-Moore) [inline openmdao/components/exec_comp.py:79]: I think `requires_fd` is missing from here.

File: openmdao/components/exec_comp.py [hunk 3]
```diff
@@ -222,28 +205,110 @@ def __init__(self, exprs=[], **kwargs):
...
+                for o in onames:
+                    self._requires_fd[o] = names
+
+        if self._requires_fd:
+            inps = []
+            for out, (rhsvars, funcs) in self._requires_fd.items():
+                iset = rhsvars.difference(outs)
+                self._requires_fd[out] = (iset, funcs)
+                inps.extend(iset)
+            self._no_check_partials = False
```
Comment (Kenneth-T-Moore) [inline openmdao/components/exec_comp.py:310]: I wonder if `_no_check_partials` should be set to False even if  complex_safe was set to True for everything that was registered, in case they are mistaken about it. Might be the only way they can find out if it's some external thing that squelches the casting warnings.

File: openmdao/components/tests/test_exec_comp.py [hunk 4]
```diff
@@ -1196,6 +1218,320 @@ def test_list_outputs_resids_tol(self):
...
+            size = 10
+            om.ExecComp.register('area', lambda x: x**2, complex_safe=True)
+            p = om.Problem()
+            p.model.add_subsystem('comp', om.ExecComp('area_square = area(x)', shape=size, has_diag_partials=True))
+            p.setup()
+            p['comp.x'] = 3.
+            p.run_model()
+            assert_near_equal(p['comp.area_square'], np.ones(size) * 9., 1e-6)
+            J = p.compute_totals(of=['comp.area_square'], wrt=['comp.x'])
+            assert_near_equal(J['comp.area_square', 'comp.x'], np.eye(size) * 6., 1e-6)
```
Comment (Kenneth-T-Moore) [inline openmdao/components/tests/test_exec_comp.py:1410]: I'm not sure what this test does. It's not showing that has_diag_partials is doing anything different (and you'd only be able to tell by looking at the size of rows/cols/val for that subjac anyway).  I guess maybe it just proves that the registered function doesn't break the code that runs with has_diag_partials.

File: openmdao/components/exec_comp.py [hunk 5]
```diff
@@ -222,28 +205,110 @@ def __init__(self, exprs=[], **kwargs):
...
+                                  ' individually.')
+
+        self.options.declare('shape_by_conn', types=bool, default=False,
+                             desc='If True, shape all inputs and outputs based on their '
+                                  'connection. Default is False.')
+
+    @classmethod
+    def register(cls, name, callable_obj, complex_safe):
+        """
+        Register a callable to be used within an ExecComp expression.
```
Comment (Kenneth-T-Moore) [inline openmdao/components/exec_comp.py:238]: maybe reword to something like: "Register a callable that can be used within all ExecComp expressions."

File: openmdao/components/tests/test_exec_comp.py [hunk 6]
```diff
@@ -1196,6 +1218,320 @@ def test_list_outputs_resids_tol(self):
...
+        # because our function is complex unsafe, we must declare that the partials
+        # with respect to 'x' use 'fd' instead of 'cs'
+        comp.declare_partials('*', 'x', method='fd')
+
+        p.setup()
+        p.run_model()
+        J = p.compute_totals(of=['comp.y'], wrt=['comp.x'])
+        assert_near_equal(J['comp.y', 'comp.x'][0][0], 4., 1e-5)
+
+    def test_register_simple(self):
```
Comment (Kenneth-T-Moore) [inline openmdao/components/tests/test_exec_comp.py:1258]: I think one test that is missing is one where you have two functions in a list, but only one of them uses a registered function, and we declare fd for that one (maybe it is not complex safe).  The other function should (I think) be computed with the internal cs, so the test should have a tighter tolerance for that one.

File: openmdao/components/tests/test_exec_comp.py [hunk 7]
```diff
@@ -1196,6 +1218,320 @@ def test_list_outputs_resids_tol(self):
...
+            self.assertEqual(J['comp.area_square', 'comp.x'], 6.)
+
+    def test_register_simple_arr(self):
+        with _temporary_expr_dict():
+            size = 10
+            om.ExecComp.register('area', lambda x: x**2, complex_safe=True)
+            p = om.Problem()
+            p.model.add_subsystem('comp', om.ExecComp('area_square = area(x)', shape=size))
+            p.setup()
+            p['comp.x'] = 3.
```
Comment (Kenneth-T-Moore) [inline openmdao/components/tests/test_exec_comp.py:1277]: Using the same value for all of the vector inputs might hide a problem.

**Review by Kenneth-T-Moore — CHANGES_REQUESTED** [review]

File: openmdao/components/tests/test_exec_comp.py (hunk 1 above)
Comment (naylor-b) [reply openmdao/components/tests/test_exec_comp.py]: done

File: openmdao/components/exec_comp.py (hunk 2 above)
Comment (naylor-b) [reply openmdao/components/exec_comp.py:79]: fixed

File: openmdao/components/tests/test_exec_comp.py (hunk 4 above)
Comment (naylor-b) [reply openmdao/components/tests/test_exec_comp.py:1430]: fixed

File: openmdao/components/exec_comp.py (hunk 5 above)
Comment (naylor-b) [reply openmdao/components/exec_comp.py]: reworded

File: openmdao/components/tests/test_exec_comp.py (hunk 7 above)
Comment (naylor-b) [reply openmdao/components/tests/test_exec_comp.py]: fixed

File: openmdao/components/tests/test_exec_comp.py (hunk 6 above)
Comment (naylor-b) [reply openmdao/components/tests/test_exec_comp.py:1268]: using manual declare_partials is an all or nothing thing.  Can't mix internal cs with openmdao cs.

**Review by Kenneth-T-Moore — APPROVED** [review]
